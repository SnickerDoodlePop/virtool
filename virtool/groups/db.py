"""
Database utilities for groups.

"""

import asyncio
from asyncio import gather
from typing import List, Optional

from motor.motor_asyncio import AsyncIOMotorClientSession
from virtool_core.models.group import Group, Permissions
from virtool_core.models.user import UserNested
from virtool.groups.pg import SQLGroup
from virtool.pg.utils import get_row, get_row_by_id

import virtool.users.db
import virtool.utils
from virtool.groups.utils import merge_group_permissions
from virtool.utils import base_processor


async def create_group_from_row(mongo, pg_group) -> Group:
    return Group(
        id=pg_group.legacy_id,
        name=pg_group.name,
        permissions=Permissions(**pg_group.permissions),
        users=await fetch_group_users(mongo, pg_group.legacy_id),
    )


async def fetch_group_from_mongo(mongo, group_id) -> Optional[Group]:
    document, users = await gather(
        mongo.groups.find_one({"_id": group_id}), fetch_group_users(mongo, group_id)
    )

    if document:
        return Group(**base_processor(document), users=users)

    return None


async def fetch_complete_group(mongo, pg, group_id: str | int) -> Optional[Group]:
    """
    Search Mongo and Postgres for a Group by its ID.

    :param mongo: The MongoDB client
    :param pg: PostgreSQL AsyncEngine object
    """
    if type(group_id) is int:
        pg_group = await get_row_by_id(pg, SQLGroup, group_id)

        if pg_group:
            return await create_group_from_row(mongo, pg_group)

        return None

    pg_group, mongo_group = await asyncio.gather(
        get_row(pg, SQLGroup, ("legacy_id", group_id)),
        fetch_group_from_mongo(mongo, group_id),
    )

    if pg_group:
        return await create_group_from_row(mongo, pg_group)

    if mongo_group:
        return mongo_group

    return None


async def get_merged_permissions(mongo, id_list: List[str]) -> dict:
    """
    Get the merged permissions that are inherited as a result of membership in the groups defined in `id_list`.

    :param mongo: The MongoDB client
    :param id_list: a list of group ids
    :return: the merged permissions

    """
    groups = await asyncio.shield(
        mongo.groups.find({"_id": {"$in": id_list}}, {"_id": False}).to_list(None)
    )
    return merge_group_permissions(groups)


async def update_member_users(
    mongo,
    group_id: str,
    remove: bool = False,
    session: Optional[AsyncIOMotorClientSession] = None,
):
    groups = await mongo.groups.find({}, session=session).to_list(None)

    async for user in mongo.users.find(
        {"groups": group_id},
        ["administrator", "groups", "permissions", "primary_group"],
        session=session,
    ):
        if remove:
            user["groups"].remove(group_id)

        update_dict = {}

        if remove:
            update_dict["$pull"] = {"groups": group_id}

            if user["primary_group"] == group_id:
                update_dict["$set"]["primary_group"] = ""

        if update_dict:
            await mongo.users.find_one_and_update(
                {"_id": user["_id"]},
                update_dict,
                projection=["groups", "permissions"],
                session=session,
            )

        permissions = merge_group_permissions(
            [group for group in groups if group["_id"] in user["groups"]]
        )

        await virtool.users.db.update_keys(
            mongo,
            user["administrator"],
            user["_id"],
            user["groups"],
            permissions,
            session=session,
        )


async def fetch_group_users(mongo, group_id: str) -> List[UserNested]:
    return [
        UserNested(**base_processor(user))
        async for user in mongo.users.find({"groups": group_id})
    ]


def lookup_group_minimal_by_id(
    local_field: str = "group", set_as: str = "group"
) -> List[dict]:
    """
    Return a list of aggregation pipeline stages to lookup a group by its id and return
    only the ``_id`` and ``name`` fields.

    """
    return [
        {
            "$lookup": {
                "from": "groups",
                "let": {"group_id": f"${local_field}"},
                "pipeline": [
                    {"$match": {"$expr": {"$eq": ["$_id", "$$group_id"]}}},
                    {"$project": {"_id": False, "id": "$_id", "name": True}},
                ],
                "as": set_as,
            },
        },
        {"$set": {set_as: {"$first": f"${set_as}"}}},
    ]


def lookup_groups_minimal_by_id(
    local_field: str = "groups",
    set_as: str = "groups",
) -> List[dict]:
    """
    Return a list of aggregation pipeline stages to lookup a group by its id and return
    only the ``_id`` and ``name`` fields.

    """
    return [
        {
            "$lookup": {
                "from": "groups",
                "let": {"group_ids": f"${local_field}"},
                "pipeline": [
                    {"$match": {"$expr": {"$in": ["$_id", "$$group_ids"]}}},
                    {"$project": {"_id": False, "id": "$_id", "name": True}},
                    {"$sort": {"name": 1}},
                ],
                "as": set_as,
            },
        },
    ]
