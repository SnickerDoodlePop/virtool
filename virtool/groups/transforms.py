from typing import Any, Dict, List, Optional, Union
from virtool.mongo.core import Mongo
from virtool.data.transforms import AbstractTransform
from virtool.types import Document
from virtool.utils import base_processor
from virtool_core.models.group import GroupMinimal


class AttachGroupTransform(AbstractTransform):
    def __init__(self, mongo: Mongo, ignore_errors: bool = False):
        self._mongo = mongo
        self._ignore_errors = ignore_errors

    def _extract_primary_group_id(self, document: Document) -> Optional[str]:
        try:
            primary_group = document["primary_group"]

        except KeyError:
            if not self._ignore_errors:
                raise KeyError("Document needs a value at group or group.id")

            return None

        try:
            return primary_group["id"]
        except TypeError:
            if isinstance(primary_group, str):
                return primary_group

            return None

    async def prepare_one(self, document: Document) -> Optional[GroupMinimal]:
        group_id = self._extract_primary_group_id(document)

        group_data = base_processor(await self._mongo.groups.find_one(group_id))

        return group_data

    async def attach_one(self, document: Document, prepared: Any):
        if not prepared:
            return {**document, "primary_group": None}

        try:
            group_data = document["primary_group"]
        except KeyError:
            if self._ignore_errors:
                return document

            raise

        if isinstance(group_data, str):
            return {
                **document,
                "primary_group": GroupMinimal(**prepared),
            }

        return {**document, "primary_group": GroupMinimal(**prepared)}

    async def prepare_many(
        self, documents: List[Document]
    ) -> Dict[Union[int, str], Any]:
        group_ids = list(
            {self._extract_primary_group_id(document) for document in documents}
        )

        return {
            document["_id"]: base_processor(document)
            async for document in self._mongo.groups.find({"_id": {"$in": group_ids}})
        }

    async def attach_many(
        self, documents: List[Document], prepared: Dict[int, Any]
    ) -> List[Document]:
        return [
            await self.attach_one(
                document, prepared.get(self._extract_primary_group_id(document))
            )
            for document in documents
        ]


class AttachGroupsTransform(AbstractTransform):
    def __init__(self, mongo: Mongo, ignore_errors: bool = False):
        self._mongo = mongo
        self._ignore_errors = ignore_errors

    def _extract_group_ids(self, document: Document) -> List[str]:
        try:
            return document["groups"]

        except (KeyError, ValueError):
            return None

    async def prepare_one(self, document: Document) -> Dict[str, Dict]:
        group_ids = self._extract_group_ids(document)

        to_return = [
            base_processor(group_doc)
            async for group_doc in self._mongo.groups.find({"_id": {"$in": group_ids}})
        ]

        return to_return

    async def attach_one(self, document: Document, prepared: Any):
        to_return = {**document, "groups": [GroupMinimal(**doc) for doc in prepared]}

        return to_return

    async def prepare_many(self, documents: List[Document]):
        prepared = dict()

        for document in documents:
            prepared[base_processor(document).get("id")] = await self.prepare_one(
                document
            )

        return prepared

    async def attach_many(self, documents: List[Document], prepared: Dict[str, Any]):
        return [
            await self.attach_one(
                document, prepared.get(base_processor(document).get("id"))
            )
            for document in documents
        ]
