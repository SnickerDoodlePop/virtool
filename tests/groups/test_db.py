from virtool.groups.transforms import AttachPrimaryGroupTransform, AttachGroupsTransform
from virtool.data.transforms import apply_transforms


class TestAttachGroupTransform:
    async def test_primary_group_missing(self, snapshot, mongo):
        doc = {"id": "doc_0", "primary_group": None}

        complete_doc = await apply_transforms(doc, [AttachPrimaryGroupTransform(mongo)])

        assert complete_doc == snapshot

    async def test_primary_group_present(self, snapshot, fake2, mongo):
        doc = {"id": "doc_0", "primary_group": (await fake2.groups.create()).id}

        complete_doc = await apply_transforms(doc, [AttachPrimaryGroupTransform(mongo)])

        assert complete_doc == snapshot

    async def test_multiple_documents(self, snapshot, fake2, mongo):
        docs = [
            {
                "id": "doc_0",
                "primary_group": (await fake2.groups.create()).id,
            },
            {
                "id": "doc_1",
                "primary_group": (await fake2.groups.create()).id,
            },
            {
                "id": "doc_2",
                "primary_group": (await fake2.groups.create()).id,
            },
        ]

        complete_docs = await apply_transforms(
            docs, [AttachPrimaryGroupTransform(mongo)]
        )

        assert complete_docs == snapshot


class TestAttachGroupsTransform:
    async def test_groups_missing(self, snapshot, mongo):
        doc = {"id": "doc", "groups": []}

        complete_doc = await apply_transforms(doc, [AttachGroupsTransform(mongo)])

        assert complete_doc == snapshot

    async def test_groups_present(self, snapshot, fake2, mongo):
        doc = {
            "id": "doc",
            "groups": [
                (await fake2.groups.create()).id,
                (await fake2.groups.create()).id,
            ],
        }

        complete_doc = await apply_transforms(doc, [AttachGroupsTransform(mongo)])

        assert complete_doc == snapshot

    async def test_multiple_documents(self, snapshot, fake2, mongo):
        docs = [
            {
                "id": "doc_0",
                "groups": [
                    (await fake2.groups.create()).id,
                    (await fake2.groups.create()).id,
                ],
            },
            {
                "id": "doc_1",
                "groups": [
                    (await fake2.groups.create()).id,
                    (await fake2.groups.create()).id,
                ],
            },
            {
                "id": "doc_2",
                "groups": [
                    (await fake2.groups.create()).id,
                    (await fake2.groups.create()).id,
                ],
            },
        ]

        complete_docs = await apply_transforms(docs, [AttachGroupsTransform(mongo)])

        assert complete_docs == snapshot
