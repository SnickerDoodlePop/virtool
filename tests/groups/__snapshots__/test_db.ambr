# name: TestAttachGroups.test_multiple_documents
  <class 'list'> [
    <class 'dict'> {
      'active': True,
      'administrator': False,
      'force_reset': False,
      'groups': <class 'list'> [
        <class 'dict'> {
          'id': 'bf1b993c',
          'name': 'musicians',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
        <class 'dict'> {
          'id': 'fb085f7f',
          'name': 'hydrogeologists',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
        <class 'dict'> {
          'id': '7cf872dc',
          'name': 'architects',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
      ],
      'handle': 'kelsey56',
      'id': '3cbb22cc',
      'invalidate_sessions': False,
      'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 32, 127000),
      'password': b'$2b$12$eMukN16d32Epzpn19NoGKehwY.dRyqxno6hLk.PqA3v9yRBZ3KCq.',
      'primary_group': None,
      'settings': <class 'dict'> {
        'quick_analyze_workflow': 'pathoscope_bowtie',
        'show_ids': True,
        'show_versions': True,
        'skip_quick_analyze_dialog': True,
      },
    },
    <class 'dict'> {
      'active': True,
      'administrator': False,
      'force_reset': False,
      'groups': <class 'list'> [
        <class 'dict'> {
          'id': 'da42dca3',
          'name': 'neurosurgeons',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
        <class 'dict'> {
          'id': '09d01419',
          'name': 'optometrists',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
        <class 'dict'> {
          'id': 'f60bae36',
          'name': 'haematologists',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
      ],
      'handle': 'darrellmorris',
      'id': '153ff459',
      'invalidate_sessions': False,
      'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 32, 436000),
      'password': b'$2b$12$H1ixiHoHwjIogAAWmDrj6.QjjuGoftWw9o9Vv.LtpJis2aSzaGLc.',
      'primary_group': None,
      'settings': <class 'dict'> {
        'quick_analyze_workflow': 'pathoscope_bowtie',
        'show_ids': True,
        'show_versions': True,
        'skip_quick_analyze_dialog': True,
      },
    },
  ]
---
# name: TestAttachGroups.test_no_groups
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'invalidate_sessions': False,
    'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 30, 871000),
    'password': b'$2b$12$x90itBSuW1uqXzbKuahQo.ZBLkRtmF5izM2u4ld9mWCq0dKMmzokC',
    'primary_group': None,
    'settings': <class 'dict'> {
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    },
  }
---
# name: TestAttachGroups.test_single_document
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'force_reset': False,
    'groups': <class 'list'> [
      <class 'dict'> {
        'id': '7cf872dc',
        'name': 'architects',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
      <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
      <class 'dict'> {
        'id': 'fb085f7f',
        'name': 'hydrogeologists',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
    ],
    'handle': 'kelsey56',
    'id': '3cbb22cc',
    'invalidate_sessions': False,
    'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 31, 530000),
    'password': b'$2b$12$Ir8lgiMAPI5wXPBbNuZeKOknZEuihcA57/24I02tgT.KYmY1F8DyS',
    'primary_group': None,
    'settings': <class 'dict'> {
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    },
  }
---
# name: TestAttachPrimaryGroup.test_multiple_documents
  <class 'list'> [
    <class 'dict'> {
      'active': True,
      'administrator': False,
      'force_reset': False,
      'groups': <class 'list'> [
        <class 'dict'> {
          'id': 'bf1b993c',
          'name': 'musicians',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
      ],
      'handle': 'qtaylor',
      'id': '7cf872dc',
      'invalidate_sessions': False,
      'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 30, 674000),
      'password': b'$2b$12$AVQe33DcBQ0sJEvMuqj1PuJncIZ2Mter3OKKBcWsPBjG0SgyxYVme',
      'primary_group': <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
      'settings': <class 'dict'> {
        'quick_analyze_workflow': 'pathoscope_bowtie',
        'show_ids': True,
        'show_versions': True,
        'skip_quick_analyze_dialog': True,
      },
    },
    <class 'dict'> {
      'active': True,
      'administrator': False,
      'force_reset': False,
      'groups': <class 'list'> [
        <class 'dict'> {
          'id': 'fb085f7f',
          'name': 'hydrogeologists',
          'permissions': <class 'dict'> {
            'cancel_job': False,
            'create_ref': False,
            'create_sample': False,
            'modify_hmm': False,
            'modify_subtraction': False,
            'remove_file': False,
            'remove_job': False,
            'upload_file': False,
          },
        },
      ],
      'handle': 'fcarpenter',
      'id': '3cbb22cc',
      'invalidate_sessions': False,
      'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 30, 332000),
      'password': b'$2b$12$pJCVkYZ35h3DLKJ4JAws6OJZvX2achzpO/g8tUcVUZbBx0jpUmz6G',
      'primary_group': <class 'dict'> {
        'id': 'fb085f7f',
        'name': 'hydrogeologists',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
      'settings': <class 'dict'> {
        'quick_analyze_workflow': 'pathoscope_bowtie',
        'show_ids': True,
        'show_versions': True,
        'skip_quick_analyze_dialog': True,
      },
    },
  ]
---
# name: TestAttachPrimaryGroup.test_no_primary_group
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'force_reset': False,
    'groups': <class 'list'> [
    ],
    'handle': 'leeashley',
    'id': 'bf1b993c',
    'invalidate_sessions': False,
    'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 28, 845000),
    'password': b'$2b$12$iXU1khDKmyyuwNHd5yPBTOty/yMYC27DvtOfmFuE6AaGLyQF6GMLC',
    'primary_group': None,
    'settings': <class 'dict'> {
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    },
  }
---
# name: TestAttachPrimaryGroup.test_single_document
  <class 'dict'> {
    'active': True,
    'administrator': False,
    'force_reset': False,
    'groups': <class 'list'> [
      <class 'dict'> {
        'id': 'bf1b993c',
        'name': 'musicians',
        'permissions': <class 'dict'> {
          'cancel_job': False,
          'create_ref': False,
          'create_sample': False,
          'modify_hmm': False,
          'modify_subtraction': False,
          'remove_file': False,
          'remove_job': False,
          'upload_file': False,
        },
      },
    ],
    'handle': 'laura56',
    'id': 'fb085f7f',
    'invalidate_sessions': False,
    'last_password_change': datetime.datetime(2023, 8, 23, 20, 34, 29, 469000),
    'password': b'$2b$12$/94GLS5pHg9flLL41.Uqf.4ZYcqDOYLaPL2jKAa3b0vUn6WCnci6e',
    'primary_group': <class 'dict'> {
      'id': 'bf1b993c',
      'name': 'musicians',
      'permissions': <class 'dict'> {
        'cancel_job': False,
        'create_ref': False,
        'create_sample': False,
        'modify_hmm': False,
        'modify_subtraction': False,
        'remove_file': False,
        'remove_job': False,
        'upload_file': False,
      },
    },
    'settings': <class 'dict'> {
      'quick_analyze_workflow': 'pathoscope_bowtie',
      'show_ids': True,
      'show_versions': True,
      'skip_quick_analyze_dialog': True,
    },
  }
---
